bart/sched/pelt.py

```diff
@@ -0,0 +1,347 @@
+#    Copyright 2015-2016 ARM Limited
+#    Copyright 2016 Google Inc. All Rights Reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+PELT Simulation Model
+"""
+
+from collections import namedtuple as _namedtuple
+from pandas import DataFrame as _DataFrame
+
+################################################################################
+# Task Model
+################################################################################
+
+class Task(object):
+
+    pelt_sample_us = 1024
+    pelt_max = 1024
+
+    def __init__(self, period_ps, run_ps, start_ps=0,
+                 pelt_sample_us=1024, pelt_max=1024):
+        """
+        A task which activations are specified in terms
+        of PELT samples.
+        A PELT sample is 1.024 ms long, this is the time
+        interval used by the Linux scheduler to update a
+        PELT signal.
+
+        :param start_ps: task start time
+        :type start_ps: int
+
+        :param period_ps: the period of each activation
+        :type period_ps: int
+
+        :param run_ps: the duration of each activation
+        :type run_ps: int
+        """
+        self.start_us  = Task.pelt_sample_us * start_ps
+        self.period_us = Task.pelt_sample_us * period_ps
+        self.run_us    = Task.pelt_sample_us * run_ps
+
+        self.pelt_sample_us = pelt_sample_us
+        self.pelt_max = pelt_max
+
+        self.idle_us   = self.period_us - self.run_us
+        self.duty_cycle_pct = int(100. * run_ps / period_ps)
+        self.util_avg =  int(float(Task.pelt_max) * self.duty_cycle_pct / 100)
+
+    def running(self, time_ms):
```

Comment:
We need to extend this method to provide a more generic implementation where the input waveform is not necessary periodic. Moreover, we should return the "percentage" of time spent since the last PELT sample as a base for supporting more generic (i.e. non periodic tasks)